sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/ObjectsShouldBeInitializedCorrectlyBase.cs

```diff
@@ -0,0 +1,154 @@
+﻿/*
+ * SonarAnalyzer for .NET
+ * Copyright (C) 2015-2018 SonarSource SA
+ * mailto: contact AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules
+{
+    public abstract class ObjectsShouldBeInitializedCorrectlyBase<TExpectedValueType> : SonarDiagnosticAnalyzer
+    {
+        protected abstract DiagnosticDescriptor Rule { get; }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        internal abstract KnownType TrackedType { get; }
+
+        protected abstract string TrackedPropertyName { get; }
+
+        protected abstract TExpectedValueType ExpectedPropertyValue { get; }
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var objectCreation = (ObjectCreationExpressionSyntax)c.Node;
+
+                    if (IsTrackedType(objectCreation.Type, c.SemanticModel) &&
+                        !IsInitializedAsExpected(objectCreation, c.SemanticModel) &&
+                        !IsLaterAssignedWithExpectedValue(objectCreation, c.SemanticModel))
+                    {
+                        c.ReportDiagnosticWhenActive(Diagnostic.Create(Rule, objectCreation.GetLocation()));
+                    }
+                },
+                SyntaxKind.ObjectCreationExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(
+                c =>
+                {
+                    var assignment = (AssignmentExpressionSyntax)c.Node;
+
+                    // Ignore assignments within object initializers, they are
+                    // reported in the ObjectCreationExpression handler
+                    if (assignment.FirstAncestorOrSelf<InitializerExpressionSyntax>() == null &&
+                        IsTrackedPropertyName(assignment.Left) &&
```

Comment:
Minor: Iit might be more efficient to put the _IsTrackerPropertyName_ check before the _FirstAncestorOrSelf_ check.

Reply:
I was considering that, but the difference would be minimal imo and also the comment placement would have been awkward (it is too long to be on the same line)